toolchain/check/impl_validation.cpp

```diff
@@ -77,6 +77,91 @@ static auto IsSameLibrary(Context& context, SemIR::InstId owning_inst_id)
 }
 
+// Determines whether the given generic parameter type can have a value that
+// introduces additional associated libraries during impl lookup.
+static auto TypeCanIntroduceAssociatedLibraries(Context& context,
+                                                SemIR::TypeId initial_type_id)
+    -> bool {
+  llvm::SmallVector<SemIR::TypeId> worklist = {initial_type_id};
+  auto push_type_inst_id = [&](SemIR::TypeInstId type_inst_id) {
+    worklist.push_back(context.types().GetTypeIdForTypeInstId(type_inst_id));
+  };
+
+  while (!worklist.empty()) {
+    auto type_id = worklist.pop_back_val();
+
+    // Remove type qualifiers and look through adapters.
+    type_id = context.types().GetObjectRepr(type_id);
+    if (!type_id.has_value()) {
+      // TODO: How should we handle the case where the parameter type is
+      // incomplete? Can we defer determining whether an impl is effectively
+      // final until we see its definition? It's not clear that a parameter of
+      // incomplete type can be used in an impl declaration, so this may not
+      // matter. For now we treat such impls as not effectively final.
+      return false;
+    }
+
+    CARBON_KIND_SWITCH(context.types().GetAsInst(type_id)) {
+      case CARBON_KIND(SemIR::ArrayType inst): {
+        push_type_inst_id(inst.element_type_inst_id);
+        break;
+      }
+      case CARBON_KIND(SemIR::StructType inst): {
+        for (auto field : context.struct_type_fields().Get(inst.fields_id)) {
+          push_type_inst_id(field.type_inst_id);
+        }
+        break;
+      }
+      case CARBON_KIND(SemIR::TupleType inst): {
+        for (auto inst_id : context.inst_blocks().Get(inst.type_elements_id)) {
+          push_type_inst_id(context.types().GetAsTypeInstId(inst_id));
+        }
+        break;
+      }
+      case SemIR::BoolType::Kind:
+      case SemIR::CharLiteralType::Kind:
+      case SemIR::FloatLiteralType::Kind:
+      case SemIR::FloatType::Kind:
+      case SemIR::IntLiteralType::Kind:
+      case SemIR::IntType::Kind: {
+        // Values of these types cannot introduce any associated library
+        // dependence.
+        return false;
+      }
+      default: {
+        // Conservatively assume any other type can introduce associated
+        // libraries.
+        return true;
+      }
+    }
+  }
+  return false;
+}
+
+// Returns whether the given `impl` is specializable in libraries other than the
+// one in which it's defined. If not, we allow it to be declared as `final`.
+static auto IsImplExternallySpecializable(Context& context,
+                                          const SemIR::Impl& impl) -> bool {
+  // If the impl is not generic, it can't be specialized.
+  if (!impl.generic_id.has_value()) {
+    return false;
+  }
+
+  // If the types of all the generic bindings for the impl do not allow values
+  // with associated libraries, then no further specialization is possible in
+  // any other library, so the impl is effectively final.
```

Comment:
I think this comment might need to be updated. It's not wrong but it's talking about the old context of this PR, but here we're checking for an impl being externally specializable, a somewhat inverse of effectively final